exporter/kinesisexporter/exporter.go

```diff
@@ -16,57 +16,90 @@ package kinesisexporter
 	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/consumer/pdata"
 	jaegertranslator "go.opentelemetry.io/collector/translator/trace/jaeger"
 	"go.uber.org/zap"
 )
 
-// Exporter implements an OpenTelemetry trace exporter that exports all spans to AWS Kinesis
+const (
+	errInvalidContext = "invalid context"
+)
+
+// Exporter implements an OpenTelemetry exporter that pushes OpenTelemetry data to AWS Kinesis
 type Exporter struct {
-	kinesis *kinesis.Exporter
-	logger  *zap.Logger
+	producer producer
+	logger   *zap.Logger
 }
 
-var _ component.TraceExporter = (*Exporter)(nil)
+// newExporter creates a new exporter with the passed in configurations.
+// It starts the AWS session and setups the relevant connections.
+func newExporter(c *Config, logger *zap.Logger) (*Exporter, error) {
+	pr, err := newKinesisProducer(c, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Exporter{producer: pr, logger: logger}, nil
+}
 
 // Start tells the exporter to start. The exporter may prepare for exporting
 // by connecting to the endpoint. Host parameter can be used for communicating
-// with the host after Start() has already returned. If error is returned by
-// Start() then the collector startup will be aborted.
-func (e Exporter) Start(_ context.Context, _ component.Host) error {
+// with the host after start() has already returned. If error is returned by
+// start() then the collector startup will be aborted.
+func (e *Exporter) Start(ctx context.Context, _ component.Host) error {
+	if ctx == nil || ctx.Err() != nil {
+		return fmt.Errorf(errInvalidContext)
+	}
+
+	e.producer.start()
 	return nil
 }
 
-// Shutdown is invoked during exporter shutdown.
-func (e Exporter) Shutdown(context.Context) error {
-	e.kinesis.Flush()
+// Shutdown is invoked during exporter shutdown
+func (e *Exporter) Shutdown(ctx context.Context) error {
+	if ctx == nil || ctx.Err() != nil {
+		return fmt.Errorf(errInvalidContext)
+	}
+
+	e.producer.stop()
 	return nil
 }
 
-// ConsumeTraceData receives a span batch and exports it to AWS Kinesis
-func (e Exporter) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+func (e *Exporter) pushTraces(ctx context.Context, td pdata.Traces) (int, error) {
+	if ctx == nil || ctx.Err() != nil {
+		return 0, fmt.Errorf(errInvalidContext)
+	}
+
 	pBatches, err := jaegertranslator.InternalTracesToJaegerProto(td)
 	if err != nil {
 		e.logger.Error("error translating span batch", zap.Error(err))
-		return consumererror.Permanent(err)
+		return td.SpanCount(), consumererror.Permanent(err)
 	}
+
 	// TODO: Use a multi error type
 	var exportErr error
 	for _, pBatch := range pBatches {
 		for _, span := range pBatch.GetSpans() {
 			if span.Process == nil {
 				span.Process = pBatch.Process
 			}
-			err := e.kinesis.ExportSpan(span)
+
+			spanBytes, err := span.Marshal()
 			if err != nil {
+				e.logger.Error("error marshaling span to bytes", zap.Error(err))
+				exportErr = err
```

Comment:
should this case return a permanent error as it's unlikely the same batch can be marshaled successfully on a later attempt